sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2796,7 +2796,31 @@ checkDevSupportDeviceRequirements(const device &Dev,
       Dims++;
     }
 
-    if (NDRDesc.Dims != 0 && NDRDesc.Dims != static_cast<size_t>(Dims))
+    int NumDimSpecified = 0;
+    if (WorkGroupNumDim) {
+      // We know the dimensions have been padded to 3, make sure that the pad
+      // value is always set to 1 and that the non-padded dimension matches the
+      // ones specified in the ND range.
+      NumDimSpecified = static_cast<int>(
+          DeviceBinaryProperty(*(WorkGroupNumDim.value())).asUint32());
+      if (NDRDesc.Dims != 0 &&
+          NDRDesc.Dims != static_cast<size_t>(NumDimSpecified))
+        return sycl::exception(
+            sycl::errc::nd_range,
+            "The local size dimension of submitted nd_range doesn't match the "
+            "required work-group size dimension");
+      for (unsigned i = NumDimSpecified; i < 3; ++i)
+        if (ReqdWGSizeVec[i] != 1)
+          return sycl::exception(sycl::errc::nd_range,
+                                 "The local size dimension of submitted "
+                                 "nd_range doesn't match the "
+                                 "required work-group size dimension");
```

Comment:
This would be an internal bug, right? I.e. somehow the padding wasn't applied correctly, but there's no way for the user to reach this. Maybe we should do:
```suggestion
#ifdef NDEBUG
      for (unsigned i = NumDimSpecified; i < 3; ++i)
        assert(ReqdWGSizeVec[i] != 1 && "Incorrect padding in required work-group size metadata.");
#endif // NDEBUG
```